src/Language/Haskell/Liquid/Bare/DataType.hs

```diff
@@ -156,3 +158,34 @@ makeTyConEmbeds' :: TCEmb (Located Symbol) -> BareM (TCEmb TyCon)
 makeTyConEmbeds' z = M.fromList <$> mapM tx (M.toList z)
   where
     tx (c, y) = (, y) <$> lookupGhcTyCon c
+
+makeRecordSelectorSigs :: [(DataCon, Located DataConP)] -> BareM [(Var, Located SpecType)]
+makeRecordSelectorSigs dcs = concat <$> mapM makeOne dcs
+  where
+  makeOne (dc, Loc l l' dcp)
+    | null (dataConFieldLabels dc)
+    = return []
+    | otherwise = do
+        fs <- mapM lookupGhcVar (dataConFieldLabels dc)
+        return (fs `zip` ts)
+    where
+    ts   = [ Loc l l' (mkArrow (freeTyVars dcp) [] (freeLabels dcp)
+                               [(z, res, mempty)]
+                               (dropPreds (subst su t `strengthen` mt)))
+           | (x, t) <- reverse args -- NOTE: the reverse here is correct
+           , not (isFunTy t) -- NOTE: we only have measures for non-function fields
+           , let vv = rTypeValueVar t
+             -- the measure singleton refinement, eg `v = getBar foo`
+           , let mt = uReft (vv, PAtom Eq (EVar vv) (EApp (EVar x) (EVar z)))
+           ]
+
+    su   = mkSubst $ [ (x, EApp (EVar x) (EVar z)) | x <- xs ]
+    args = tyArgs dcp
+    xs   = map fst args
+    z    = "lq$recSel"
+    res  = dropPreds (tyRes dcp)
+
+    -- FIXME: this is clearly imprecise, but the preds in the DataConP seem
+    -- to be malformed. If we leave them in, tests/pos/kmp.hs fails with
+    -- a malformed predicate application. Niki, help!!
+    dropPreds = fmap (\(MkUReft r _ps ss) -> MkUReft r mempty ss)
```

Comment:
@nikivazou I'm not sure what's going on here, but we seem to create a malformed predicate application in datacon types, specifically in the result type.

It was already there prior to this patch (look at the type for the datacon in `tests/pos/kmp.hs`), but somehow it hasn't caused problems so far.